### PR TITLE
Add tool chain for Netcore Mono AOT.

### PR DESCRIPTION
**Pull request by naricc**
This adds a tool chain for doing AOT with Mono using Netcore (its a little different than classic Mono).

Directions for building an appropriate runtime and using the tool chain: https://gist.github.com/naricc/f99b80ea5ec7f4a8f4b4777cd46ef7b8

`--AOTCompilerMode llvm`

Is not working right now because of a bug in dotnet/runtime, but as far as I can tell everything works with it on the BDN side.

### CODE REVIEW

**Comment by naricc** [comment]
@adamsitnik

**Comment by adamsitnik** [comment]
> Directions for building an appropriate runtime and using the tool chain: 

Please add them to this file: https://github.com/dotnet/BenchmarkDotNet/blob/master/docs/articles/configs/toolchains.md#wasm

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:81]: we already have such a method, please reuse it: 

https://github.com/dotnet/BenchmarkDotNet/blob/502dc9aca86f21a9d38da9634f3bec2143f1f6f5/src/BenchmarkDotNet/Toolchains/DotNetCli/CustomDotNetCliToolchainBuilder.cs#L136-L144

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs:33]: please remove the extra empty lines and fix the indentation in this file

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:86]: the new builder calls `dotnet build`, not `dotnet publish` (`dotnetCliBuilder.Build`). Are you sure that the output will be in the publish folder?

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMBuilder.cs
Comment (vargaz) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMBuilder.cs:61]: Extra logging.

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs
Comment (vargaz) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs:8]: Extra lines.

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs
Comment (vargaz) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs:22]: Indentation

File: src/BenchmarkDotNet/Templates/MonoAOTLLVMCsProj.txt
Comment (vargaz) [inline src/BenchmarkDotNet/Templates/MonoAOTLLVMCsProj.txt:45]: Extra lines.

File: src/BenchmarkDotNet/Templates/MonoAOTLLVMCsProj.txt
Comment (adamsitnik) [inline src/BenchmarkDotNet/Templates/MonoAOTLLVMCsProj.txt:14]: We should most probably try to avoid using a hardcoded version? Should this be configurable? Or using the latest by having the asterisk?

```suggestion
    <PackageReference Include="Microsoft.NET.Runtime.MonoAOTCompiler.Task" version="6.0.0-*" GeneratePathProperty="true" />
```

File: src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:31]: You mention `RuntimeInformation.GetCurrentRuntime`, but seem to forget to include the changes in this method in this PR.

**Review by adamsitnik — COMMENTED** [review]
Overall looks good, but please address my comments before I hit the merge button.

Thank you @naricc !

File: src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs
Comment (naricc) [reply src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:31]: Actually I think I cribbed this comment from wasm; do we actually need any change to RuntimeInformation?

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMBuilder.cs
Comment (naricc) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMBuilder.cs]: Thanks, didn't realize that method existed.

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (naricc) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:83]: Since that method is protected, what is the best way to reuse it? Just make it public? It doesn't seem like I need to use the whole tool chain builder.

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (naricc) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:88]: I think since I have done this with the AotApp Target:

`<Target Name="AotApp" AfterTargets="Build" DependsOnTargets="Publish">`

It will always run after build, and force a publish.

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (adamsitnik) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:83]: You can make it `internal`

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (adamsitnik) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:88]: thanks for the explanation!

File: src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs
Comment (adamsitnik) [reply src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:31]: >  do we actually need any change to RuntimeInformation

Yes, otherwise BDN will print invalid data in the table (it will recognize the runtime as .NET 6?)

Comment (naricc) [reply src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:31]: So, to do this, I'd need to have some way for C# code to know if it is being run in AOT? I think right now it actually hits the mono case (but says nothing about AOT).

Comment (naricc) [reply src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:31]: Hmm, I can't actually find a way to detect if we are using AOT here. I could use `System.Runtime.CompilerServices.RuntimeFeature`, but its not available in netstandard2.0.

It seems like it will detect 'mono', which is the runtime, but not the fact that it is AOT.

File: docs/articles/configs/toolchains.md
Comment (adamsitnik) [inline docs/articles/configs/toolchains.md:378]: When running benchmarks we should always be using `Release` builds.

```suggestion
`./build.sh -subset mono+libs -c Release`
```

File: docs/articles/configs/toolchains.md
Comment (adamsitnik) [inline docs/articles/configs/toolchains.md:383]:
```suggestion
<runtime root>/artifacts/obj/mono/OSX.x64.Release/mono/mini/mono-sgen
```

File: docs/articles/configs/toolchains.md
Comment (adamsitnik) [inline docs/articles/configs/toolchains.md:389]:
```suggestion
<runtimeroot>artifacts/bin/microsoft.netcore.app.runtime.osx-x64/Release/
```

File: docs/articles/configs/toolchains.md
Comment (adamsitnik) [inline docs/articles/configs/toolchains.md:392]:
```suggestion
```

File: src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs
Comment (adamsitnik) [reply src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:31]: > I could use System.Runtime.CompilerServices.RuntimeFeature, but its not available in netstandard2.0.

In such a case you can use reflection to read the data:

```cs
private static bool IsAOT()
{
    Type runtimeFeature = Type.GetType("System.Runtime.CompilerServices.RuntimeFeature");
    if (runtimeFeature != null)
    {
        MethodInfo methodInfo = runtimeFeature.GetProperty("IsDynamicCodeCompiled", BindingFlags.Public | BindingFlags.Static)?.GetMethod;
        if (methodInfo != null)
        {
            return (bool)methodInfo.Invoke(null, null);
        }
    }

    return false;
}
```

File: src/BenchmarkDotNet/Toolchains/DotNetCli/NetCoreAppSettings.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/DotNetCli/NetCoreAppSettings.cs:106]:
```suggestion
```

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotCompilerMode.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotCompilerMode.cs:1]:
```suggestion
﻿using System;
﻿
```

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMBuilder.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMBuilder.cs:27]:
```suggestion
```

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:13]:
```suggestion
```

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMGenerator.cs:41]: this can be simplified

```suggestion
            string useLLVM = AotCompilerMode == MonoAotCompilerMode.llvm ? "true" : "false";
```

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs:14]:
```suggestion
```

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs
Comment (adamsitnik) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs:33]: @naricc you seem to have forgotten to push a commit as you have marked all the comments about empty lines and indentation as resolved, while they have not been resolved. Please fix it.

**Review by adamsitnik — COMMENTED** [review]
Overall looks good, but there are still two things missing:

* the possibility to detect Mono AOT at Runtime. Please use reflection and implement it [here](https://github.com/dotnet/BenchmarkDotNet/blob/e1c8cb89607abaa84c4881e1a73ee0f1ae76a3b8/src/BenchmarkDotNet/Portability/RuntimeInformation.cs#L125-L127)
* whitespaces and indentation fixes

@naricc will it work only on Unix, or should I be able to test it on Windows as well?

File: docs/articles/configs/toolchains.md
Comment (akoeplinger) [inline docs/articles/configs/toolchains.md:368]: typo:
```suggestion
and optionally (defaults to mini)
```

File: src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs
Comment (akoeplinger) [inline src/BenchmarkDotNet/Environments/Runtimes/MonoAotLLVMRuntime.cs:20]: `mainJs` isn't used here

**Comment by adamsitnik** [comment]
I am getting the following error when I try to test it locally:

```log
error MSB4062: The "MonoAOTCompiler" task could not be loaded from the assembly C:\Users\adsitnik\.nuget\packages\microsoft.net.runtime.monoaotcompiler.task\6.0.0-preview.1.21102.12/MonoAOTCompiler.dll. Could not load file or assembly 'C:\Users\adsitnik\.nuget\packages\microsoft.net.runtime.monoaotcompiler.task\6.0.0-preview.1.21102.12\MonoAOTCompiler.dll'. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.
```

@naricc Is it known? Should I just update sth?

File: src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs
Comment (naricc) [reply src/BenchmarkDotNet/Toolchains/MonoAotLLVM/MonoAotLLVMToolChain.cs:33]: Sorry, I'm not sure how I managed to do that. I will fix it.

**Comment by naricc** [comment]
> I am getting the following error when I try to test it locally:
> 
> ```
> error MSB4062: The "MonoAOTCompiler" task could not be loaded from the assembly C:\Users\adsitnik\.nuget\packages\microsoft.net.runtime.monoaotcompiler.task\6.0.0-preview.1.21102.12/MonoAOTCompiler.dll. Could not load file or assembly 'C:\Users\adsitnik\.nuget\packages\microsoft.net.runtime.monoaotcompiler.task\6.0.0-preview.1.21102.12\MonoAOTCompiler.dll'. The system cannot find the file specified. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.
> ```
> 
> @naricc Is it known? Should I just update sth?

It should be able to find it from this line in the csproj:

`<UsingTask TaskName="MonoAOTCompiler" AssemblyFile="$(PkgMicrosoft_NET_Runtime_MonoAOTCompiler_Task)/MonoAOTCompiler.dll" /> `

It is using that package relative path feature to find the assembly for the task. Have I accidentally done something non-portable here?

**Comment by adamsitnik** [comment]
> Have I accidentally done something non-portable here?

After adding .NET 6 NuGet feed to [nuget.config](https://github.com/dotnet/BenchmarkDotNet/blob/master/NuGet.Config) file:

```xml
<add key="dotnet6" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json" />
```

this problem is gone. Now I am getting;

```log
'"clang.exe"' is not recognized as an internal or external command
```

let me try a few things and get back to you

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @naricc !

**Comment by adamsitnik** [comment]
@naricc when [this](https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/builds/38139175) build finishes (should be 30-40 minutes from now) BDN CI is going to produce `0.12.1.1520` NuGet package. To use it in the performance repo you will have to ask @billwert or @Drewscoggins to move it to our internal MS BDN NuGet feed

**Pull request merged**
